src/Kiota.Builder/KiotaBuilder.cs

```diff
@@ -1017,56 +1041,102 @@ private CodeTypeBase CreateCollectionModelDeclaration(OpenApiUrlTreeNode current
-
-        var factoryMethod = AddDiscriminatorMethod(newClass, discriminator?.PropertyName);
-        
         CreatePropertiesForModelClass(currentNode, schema, currentNamespace, newClass); // order matters since we might be recursively generating ancestors for discriminator mappings and duplicating additional data/backing store properties
-        
-        if (discriminator?.Mapping?.Any() ?? false)
-            discriminator.Mapping
-                .Where(x => !x.Key.TrimStart('#').Equals(schema.Reference?.Id, StringComparison.OrdinalIgnoreCase))
-                .Select(x => (x.Key, GetCodeTypeForMapping(currentNode, x.Value, currentNamespace, newClass, schema)))
-                .Where(x => x.Item2 != null)
-                .ToList()
-                .ForEach(x => factoryMethod.AddDiscriminatorMapping(x.Key, x.Item2));
-
+        AddDiscriminatorMethod(newClass, GetDiscriminatorPropertyName(schema), GetDiscriminatorMappings(currentNode, schema, currentNamespace, newClass));
         return newClass;
     }
-    public static CodeMethod AddDiscriminatorMethod(CodeClass newClass, string discriminatorPropertyName) {
-        var factoryMethod = newClass.AddMethod(new CodeMethod {
+    private static string GetDiscriminatorPropertyName(OpenApiSchema schema) {
+        if(schema == null)
+            return string.Empty;
+        if(schema.Discriminator?.Mapping == null)
+            if(schema.OneOf.Any())
+                return schema.OneOf.Select(static x => GetDiscriminatorPropertyName(x)).FirstOrDefault(static x => !string.IsNullOrEmpty(x));
+            else if (schema.AnyOf.Any())
+                return schema.AnyOf.Select(static x => GetDiscriminatorPropertyName(x)).FirstOrDefault(static x => !string.IsNullOrEmpty(x));
+            else if (schema.AllOf.Any())
+                return schema.AllOf.Last().Discriminator?.PropertyName;
+            else
+                return string.Empty;
+
+        return schema.Discriminator.PropertyName;
+    }
+    private static readonly Func<OpenApiSchema, bool> allOfEvaluator = static x => x.Discriminator?.Mapping.Any() ?? false;
+    private IEnumerable<KeyValuePair<string, CodeTypeBase>> GetDiscriminatorMappings(OpenApiUrlTreeNode currentNode, OpenApiSchema schema, CodeNamespace currentNamespace, CodeClass baseClass) {
+        if(schema == null)
+            return Enumerable.Empty<KeyValuePair<string, CodeTypeBase>>();
+        if(!(schema.Discriminator?.Mapping?.Any() ?? false))
+            if(schema.OneOf.Any())
+                return schema.OneOf.SelectMany(x => GetDiscriminatorMappings(currentNode, x, currentNamespace, baseClass));
+            else if (schema.AnyOf.Any())
+                return schema.AnyOf.SelectMany(x => GetDiscriminatorMappings(currentNode, x, currentNamespace, baseClass));
+            else if (schema.AllOf.Any(allOfEvaluator))
+                return GetDiscriminatorMappings(currentNode, schema.AllOf.Last(allOfEvaluator), currentNamespace, baseClass);
```

Comment:
We may need to re-evaluate this block.
If there is an `allOf` with the first item (e.g. `entity`) which matches the evaluator and the second item does not match, we will pass down schema the schema for `entity` however since we are looking for the discriminator for the `baseClass` which is not `entity` and is typically the last entry in the `AllOf` collection, we will have a corrupted CodeDom as we will pass the wrong schema object for model creation when indeed we just didn't have any discriminator info for the object.

The logic is also inconsistent with the `GetDiscriminatorPropertyName` function on line 1048 when handling a schema with an `AllOf`

Reply:
good catch, revamped the method

Reply:
Thanks! Looks good now.